synapseclient/core/credentials/cred_data.py

```diff
@@ -121,16 +125,87 @@ def __repr__(self):
+
+    Returns:
+        dict: A dictionary mapping profile names to authentication credentials:
+              {
+                  "default": {"username": "your_username", "auth_token": "your_auth_token"},
+                  "profile1": {"username": "another_username", "auth_token": "another_auth_token"},
+                  ...
+              }
+
+    Raises:
+        FileNotFoundError: If the configuration file does not exist.
+    """
+
+    # Check if the configuration file exists before attempting to read it
+    if not os.path.exists(config_path):
+        raise FileNotFoundError(f"Config file not found: {config_path}")
+
+    config = configparser.ConfigParser()
+    config.read(config_path)
+
+    auth_profiles = {}
+
+    for section in config.sections():
+        normalized_profile = section if section == "default" else section.replace("profile ", "")
+
+        auth_profiles[normalized_profile] = {
+            "username": config.get(section, "username", fallback=None),
+            "auth_token": config.get(section, "authtoken", fallback=None),
+        }
+
+    return auth_profiles
+
+
+@dataclass
+class UserLoginArgs:
+    """
+    Data class representing user login arguments for authentication.
+
+    This class encapsulates authentication credentials, supporting both profile-based
+    authentication (from the Synapse configuration file) and direct authentication
+    using an auth token.
+
+    Attributes:
+        profile (Optional[str]): The profile name to use for authentication.
+                                 Defaults to "default".
+        username (Optional[str]): The Synapse username associated with the login profile.
+                                  Defaults to None.
+        auth_token (Optional[str]): The authentication token for logging in.
+                                    Hidden from debug logs for security.
+    """
+    profile: Optional[str] = field(default="default")
+    username: Optional[str] = field(default=None)
+    auth_token: Optional[str] = field(default=None, repr=False)  # Hide auth_token from debug logs
+
+    def __post_init__(self):
+        """
+        Automatically fetch credentials from the configuration file if none are provided.
+
+        This method ensures that if a username and auth token are not explicitly supplied,
+        they will be retrieved from the `~/.synapseConfig` file based on the specified profile.
+
+        Raises:
+            ValueError: If the specified profile is not found in the configuration file.
+        """
+        if self.auth_token is None and self.username is None:
+            auth_profiles = get_config_authentication()
+            if self.profile in auth_profiles:
+                self.username = auth_profiles[self.profile]["username"]
+                self.auth_token = auth_profiles[self.profile]["auth_token"]
+            else:
+                raise ValueError(f"Profile '{self.profile}' not found in config.")
```

Comment:
I do not think that we should preemptively retrieve this data. I see that the only purpose of this `UserLoginArgs` is to hold it's 3 attributes in one place.

It's usage of `get_config_authentication` is also a concern. There are now 2 instances of this function, whereas we should only have 1.

`synapseclient/core/credentials/cred_data.py::get_config_authentication` should be removed.

`synapseclient/api/configuration_services.py::get_config_authentication` Should be the one that is kept and re-used wherever we need to look the data up.


`synapseclient/api/configuration_services.py::get_config_authentication` should only be called at the point in time when the data is needed - Not ahead of time as done here.